src/Language/Parser.php

```diff
@@ -62,65 +62,57 @@ class Parser implements ParserInterface
     protected $lexer;
 
     /**
-     * Given a GraphQL source, parses it into a Document.
-     * Throws GraphQLError if a syntax error is encountered.
+     * @param string $name
+     * @param array  $arguments
+     * @return mixed
      *
-     * @inheritdoc
-     * @throws SyntaxErrorException
-     * @throws \ReflectionException
      * @throws InvariantException
+     * @throws SyntaxErrorException
      */
-    public function parse($source, array $options = []): DocumentNode
+    public function __call(string $name, array $arguments)
     {
-        $this->lexer = $this->createLexer($source, $options);
+        $lexCallback = str_replace('parse', 'lex', $name);
+
+        if (\method_exists($this, $lexCallback)) {
+            return $this->parsePartial([$this, $lexCallback], ...$arguments);
+        }
 
-        return $this->parseDocument();
+        return $this;
     }
 
     /**
-     * Given a string containing a GraphQL value (ex. `[42]`), parse the AST for
-     * that value.
+     * Given a GraphQL source, parses it into a Document.
      * Throws GraphQLError if a syntax error is encountered.
      *
-     * This is useful within tools that operate upon GraphQL Values directly and
-     * in isolation of complete GraphQL documents.
-     *
      * @inheritdoc
      * @throws SyntaxErrorException
+     * @throws \ReflectionException
      * @throws InvariantException
      */
-    public function parseValue($source, array $options = []): NodeInterface
+    public function parse($source, array $options = []): DocumentNode
     {
         $this->lexer = $this->createLexer($source, $options);
 
-        $this->expect(TokenKindEnum::SOF);
-        $value = $this->parseValueLiteral();
-        $this->expect(TokenKindEnum::EOF);
-
-        return $value;
+        return $this->lexDocument();
     }
 
     /**
-     * Given a string containing a GraphQL Type (ex. `[Int!]`), parse the AST for
-     * that type.
-     * Throws GraphQLError if a syntax error is encountered.
-     *
-     * This is useful within tools that operate upon GraphQL Types directly and
-     * in isolation of complete GraphQL documents.
-     *
-     * @inheritdoc
-     * @throws SyntaxErrorException
+     * @param callable      $lexCallback
+     * @param Source|string $source
+     * @param array         $options
+     * @return mixed
      * @throws InvariantException
+     * @throws SyntaxErrorException
      */
-    public function parseType($source, array $options = []): TypeNodeInterface
+    protected function parsePartial(callable $lexCallback, $source, array $options = [])
     {
         $this->lexer = $this->createLexer($source, $options);
 
         $this->expect(TokenKindEnum::SOF);
-        $type = $this->parseTypeReference();
+        $node = \call_user_func($lexCallback, $source, $options);
```

Comment:
You can do `$lexCallback($source, $options)`, no need for call_user_func

Reply:
We should probably change the use of callbacks everywhere.